include/deal.II/fe/fe_interface_values.h

```diff
@@ -0,0 +1,686 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_fe_interface_values_h
+#define dealii_fe_interface_values_h
+
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_q1.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * FEInterfaceValues is a data structure to access and assemble Finite Element
+ * data on interfaces between two cells of a mesh.
+ *
+ * It provides a way to access averages, jump terms, and similar operations used
+ * in Discontinuous Galerkin methods on a face between two neighboring cells.
+ * This allows the computation of typical mesh-dependent linear or bilinear
+ * forms in a similar way as FEValues does for cells and FEFaceValues does for
+ * faces. In
+ * the literature, the faces between neighboring cells are called "inner
+ * interfaces" or "facets".
+ *
+ * Internally, this class provides an abstraction for two FEFaceValues
+ * objects (or FESubfaceValues when using adaptive refinement). The class
+ * introduces a new "interface dof index" that walks over
+ * the union of the dof indices of the two FEFaceValues objects. Helper
+ * functions allow translating between the new "interface dof index" and the
+ * corresponding "cell index" (0 for the first cell, 1 for the second cell)
+ * and "dof index" within that cell.
+ *
+ * The class is made to be used inside MeshWorker::mesh_loop. It is intended
+ * to be a low level replacement for MeshWorker and LocalIntegrators and a
+ * higher level abstraction compared to assembling face terms manually.
+ *
+ * @author Timo Heister, 2019.
+ */
+template <int dim, int spacedim = dim>
+class FEInterfaceValues
+{
+public:
+  /**
+   * Construct an FEInterfaceValues from existing FEFaceValues objects.
+   *
+   * The arguments to this functions are used to construct the internal
+   * FEFaceValues and FESubfaceValues objects for the two cells of the
+   * interface. As the arguments are copied, there are no requirements
+   * on the lifetime of the arguments.
+   */
+  FEInterfaceValues(const FEFaceValues<dim> &   fe_face_values,
+                    const FESubfaceValues<dim> &fe_subface_values);
```

Comment:
I think I still don't understand (i) what exactly is being copied, and (ii) whether the arguments have to be initialized to a particular face or subface.

Reply:
Looking at the other constructor, I suspect that all you copy if the mapping, fe, quadrature, and update flags. But if that's what it is, I'd rather this constructor goes away altogether because I find it confusing that that is the first option. (Alternatively, move it below the other constructors and say that it forwards to these by copying out this that and the other.)